include/fiction/algorithms/physical_design/on_the_fly_circuit_design_on_defective_surface.hpp

```diff
@@ -0,0 +1,216 @@
+        auto black_list = sidb_surface_analysis<sidb_skeleton_bestagon_library>(
+            lattice_tiling, params.parameterized_gate_library_parameter.defect_surface, std::make_pair(0, 0));
+
+        const mockturtle::stopwatch stop{stats.time_total};
+
+        while (!gate_level_layout.has_value())
+        {
+            // P&R with *exact* and the pre-determined blacklist
+            gate_level_layout = exact_with_blacklist<GateLyt>(mapped_network, black_list, params.exact_design_parameter,
+                                                              &stats.exact_stats);
+            if (gate_level_layout.has_value())
+            {
+                try
+                {
+                    cell_level_layout = apply_parameterized_gate_library<CellLyt, parameterized_gate_library, GateLyt,
+                                                                         parameterized_gate_library_params<CellLyt>>(
+                        *gate_level_layout, params.parameterized_gate_library_parameter);
+                }
+
+                // on-the-fly gate design was unsuccessful at a certain tile. Hence, this tile-gate pair is added to the
+                // blacklist.
+                catch (const gate_design_exception<tt, GateLyt>& e)
+                {
+                    black_list[e.which_tile()][e.which_truth_table()].push_back(e.which_port_list());
+                }
+
+                catch (const std::exception& e)
+                {
+                    std::cerr << "Caught std::exception: " << e.what() << '\n';
+                }
+            }
+
+            // P&R was unsuccessful
+            else
+            {
+                break;
+            }
+        }
+
+        // empty defect-surface is returned when no P&R solution is found
+        if (!gate_level_layout.has_value())
+        {
+            return sidb_defect_surface<CellLyt>{};
+        }
+
+        // check equivalence
+        const auto miter = mockturtle::miter<technology_network>(mapped_network, *gate_level_layout);
+        const auto eq    = mockturtle::equivalence_checking(*miter);
+
+        // empty defect-surface is returned when equivalence check is empty
+        if (!eq.has_value())
+        {
+            return sidb_defect_surface<CellLyt>{};
+        }
+
+        // in case of equality, an sidb_defect_surface with the SiDBs of the circuit is returned
+        if (*eq)
+        {
+            sidb_defect_surface<CellLyt> sidbs_and_defects{cell_level_layout};
+
+            // add defects to the circuit.
+            params.parameterized_gate_library_parameter.defect_surface.foreach_sidb_defect(
+                [&sidbs_and_defects](const auto& defect)
+                { sidbs_and_defects.assign_sidb_defect(defect.first, defect.second); });
+
+            return sidbs_and_defects;
+        }
+
+        // in case of inequality, empty defect-surface is returned
+        return sidb_defect_surface<CellLyt>{};
```

Comment:
It's nice that you're including equality checking into your circuit generator, but that should be external for a separation of concerns. In your experiment script, you may call any equivalence checker you'd like. As a physical design algorithm, this should only do the layouting.

Reply:
You are absolutely right!

Reply:
Should we then store the found ``gate_level_layout`` to ``stats``so that the equivalence can be checked afterwards?

Reply:
That's a great idea!